### PR TITLE
Feat captcha update

### PR DESCRIPTION
**Pull request by gaawa**
# Main Changes

- This update places new users to a quarantine channel first, by giving the "entrance-role".
- A moderator shall review the user and either remove the "entrance-role" or ban the user.
- When the "entrance-role" is removed, the bot automatically adds member-role to the new member.
- Users with 'manage_messages' permissions can run `!clear-entrance` command to remove all messages in "entrance-channel".
- The bot tracks his entrance message and messages from the new users by himself to remove them after that user was permitted for member-role.

## New .json config file
The new version of the bot requires an updated `.desugreet` .json config file in the following format:

```json
{
"token": "your top secret token",
"member-role": "Member-Desu",
"entrance-role": "entrance-person",
"entrance-channel-id": "channel ID of the entrance channel",
"entrance-message": "Säli {0.mention}.\nSchriib hie schnell irgend e Scheiss uf Schwitzerdütsch ine zum bewiise, dass du ke Maschine bisch.\nE Mod wird dir denn e Member-Rolle geh.",
"welcome-message": "Hallo {0.mention}, willkommen bei {1.name} \n \n Bevor du dich ins Getümmel wirfst, bitte
lies dir kurz die Regeln für unseren Server durch. \n \n Alles verstanden? Dann viel Spass auf unserem Server! \n",
"log-channel-id": "channel ID of the log channel"
}

```

# Additional Changes
This update also brings the following changes
- Added `message_content` intent as bots need them to read messages from channels.
- When logging leaving users, also print out their #descriminator to facilitate searching for the user in discords user-clitent search bar.
- handle exception when a new member blocks messages per DM. Print out (closed DM) in the log.
- Also log when a member boost or de-boost the server.
- Mainly use swiss german for log text.

### CODE REVIEW

File: desugreet/bot.py [hunk 1]
```diff
@@ -4,44 +4,109 @@
...
+        await member.add_roles(entrance_role)
     else:
         print("check your config file, the role does not exist!")
-    await member.send(welcome_msg.format(member, server))
-    await client.get_channel(log_channel_id).send("{0.mention} has joined".format(member))
 
+    entrance_track_buffer[member] = []  # create list buffer with a member obj key
+    entrance_track_buffer[member].append(await client.get_channel(entrance_channel_id).send(entrance_msg.format(member)))
+
+None
```
Comment (TypicalFence) [inline desugreet/bot.py]: Is this None intentional?

File: desugreet/bot.py [hunk 2]
```diff
@@ -4,44 +4,109 @@
...
+    entrance_role = discord.utils.get(server.roles, name=entrance_role_str)
+    if entrance_role is not None:
+        await member.add_roles(entrance_role)
     else:
         print("check your config file, the role does not exist!")
-    await member.send(welcome_msg.format(member, server))
-    await client.get_channel(log_channel_id).send("{0.mention} has joined".format(member))
 
+    entrance_track_buffer[member] = []  # create list buffer with a member obj key
+    entrance_track_buffer[member].append(await client.get_channel(entrance_channel_id).send(entrance_msg.format(member)))
```
Comment (TypicalFence) [inline desugreet/bot.py:31]: do we need a try catch for the await there? what happens on failure there? Can probably ignore tho

File: desugreet/bot.py [hunk 3]
```diff
@@ -4,44 +4,109 @@
...
+        member_role = discord.utils.get(server.roles, name=member_role_str)
+        if member_role is not None:
+            await member_after.add_roles(member_role)
+        else:
+            print("check your config file, the role does not exist!")
+        # detect when member does not accept DM
+        try:
+            await member_after.send(welcome_msg.format(member_after, server))
+        except discord.errors.Forbidden:
+            await client.get_channel(log_channel_id).send("{0.mention} (closed DM) isch cho".format(member_after))
```
Comment (TypicalFence) [inline desugreet/bot.py:53]: Closed dm?

Comment (TypicalFence) [reply desugreet/bot.py:53]: Nevermind this one

File: desugreet/bot.py [hunk 4]
```diff
@@ -4,44 +4,109 @@
...
 
 @client.event
 async def on_message(message):
     if message.author == client.user:
         return
 
+    entrance_channel = client.get_channel(entrance_channel_id)
+    if message.channel is entrance_channel:
+        if message.author in entrance_track_buffer.keys():
+            entrance_track_buffer[message.author].append(message)
```
Comment (TypicalFence) [inline desugreet/bot.py:97]: What is the point of keeping the messages in memory exactly? I don't really get the point of it.

**Review by TypicalFence — APPROVED** [review]
Thanks @gaawa for implementing this. I am fine with it being merged if its tested.

Added some comments and questions.

File: desugreet/bot.py (hunk 1 above)
Comment (gaawa) [reply desugreet/bot.py]: nope lmao

File: desugreet/bot.py (hunk 2 above)
Comment (gaawa) [reply desugreet/bot.py:31]: I don't really feel the need for it since the failure occurs on configuration error

File: desugreet/bot.py (hunk 4 above)
Comment (gaawa) [reply desugreet/bot.py:97]: I record the messages the new entrance-person writes in a list such that I can call delete() method on them.
The list associated with the new user contains the message the bot sends at him and all messages the new user writes in the entrance-channel.
When the user is accepted to Member, delete() method is called on all messages in the list.
After that I probably should remove the dictionary entry. I forgor.

Comment (TypicalFence) [reply desugreet/bot.py:97]: we should probably save that to a file as its gone should the bot crash 🤔

**Pull request merged**